Compile-Time Literal Expression Syntax

daniels220 opened the pull request:
This PhEP describes an extension to the compiler that allows arbitrary expressions to be evaluated at compile time, with the results stored in the method's literal frame. This is inspired by and will be compatible with the implementation in Dolphin Smalltalk. I refer to such expressions interchangeably as "compile-time literals", "compile-time expressions", and "optimized expressions".

estebanlm commented:
@MarcusDenker can you take a look at this?

rjsargent commented:
Be sure to include all the literals inside the compile-time expression into the enclosing methods literals.
There is nothing worse that looking for senders of a method or references to a class and not finding one that was used inside a compile-time literal.

tesonep commented:
Really Interesting
Two comments after a short scan, I have to read it more in detail:
- We have to ensure elements stored in the literals are copied on use or inmutable in all sense (if not there is chance of producing awful bugs).
- We can include the selectors / classes used in the calculated values to help the tools to find them with senders.

daniels220 commented:
@rjsargent Making sure Browse References works properly is definitely important, yes. See the note in the PHEP about including the generated CompiledCode for the optimized expression in the outer method's literal frame. I believe this will be sufficient for browse references to find literals contained therein, but if not I would definitely include the individual classes/globals/selectors separately as well.

@tesonep Good point about "in all sense"—or, well, I'm not sure if this is what you mean, but it pointed out to me that this is actually a hard problem if an expression creates a collection of generally-mutable objects (including but not limited to other collections). We can make the top-level collection immutable for sure, but the contents could still be mutated. Yet if we simply apply #isReadOnlyObject: recursively, we risk a stray reference leading us to make the entire image immutable, so that's not an option without a great deal of care.

Off the top of my head, I admit I can't think of a case where there would be a reference from a compile-time literal to an object which predates the compile-time literal—has other objects referring to it—but is not already immutable, and is not a class.

Classes are a big "exception" here, but I think that's okay. It makes total sense to me to refer to a class in a compile-time literal (mapping classes to/from short codes in a serializer, for example), caring only about its name/identity, not a problem if ivars or methods are added/removed.

The problem here is the "predates the compile-time literal"—we certainly don't want to make random existing objects immutable. If we could _detect_ them, we could make new objects immutable, and throw an error if we encounter old ones that aren't. But while this is possible at the VM level, I don't think there are any hooks that would allow it at the image level, and frankly the idea creeps me out—I wouldn't want to do something like this without a lot of additional thought.

Honestly though? I think this is a case where we're giving the user a gun, and if they _really_ want to shoot themselves in the foot, they can, but it's not very likely in practice. Making the top-level value of the compile-time literal immutable could certainly be a reasonable step to take (though we need an exception for classes for sure—and technically other problems are possible, with e.g. "flyweight" objects that may be reused rather than instantiated anew), but anything beyond that strikes me as more trouble than it's worth. Or perhaps a compiler _warning_ when encountering mutable non-class references? That way the user is proactively informed of the problem but can ignore it if they know what's going on.

privat commented:
I did not know about this proposal when I wrote https://github.com/pharo-project/pharo/pull/11717 (the PR is closed now, but the discussion can maybe be useful).

I think I understand that the idea behind the syntax `##()` is to piggyback on the existing `#` character that has a feeling of "static" or "literal" but `##()` seems quite cryptic for a niche feature. I would prefer a more (i) explicit and (ii) extendable syntax.

For instance. `##compiletime(60+60+24)` where `##` denotes the beginning of a special expression, the keyword `compiletime` is more explicit about the semantic of the special expression (a random reader can guess what could be the behavior)  and the whole syntax can be extended for future language evolution (with different keywords) ­— having `###()` to mean something else in X years is a bit dystopic :p 

I also have questions/remarks about the context/bindings.

Binding `self` on the class make sense, but the spec should state this explicitly. I assume that class variables (shared variables) are also bound the same way? It `thisContext` also bound to something?

Also, what about the context in other places? Is `##(self)` bound to the metaclass when used on a class side method?
What about `##(self)` in the workspace or on random DoIt?

Finally, and it is the main reason I closed my PR, what are the real use cases of the feature? I'm playing the devil's advocate here, but `##(self)` feels a bad hack to means "name of the current class" ; a reasonable optimizing compiler (or VM) might want to automatically precompute ##(60*60*24) anyway; and global statically allocated mutable objects like dictionaries are often considered a bad idea as they easily break modularity, isolation and unit testing.

daniels220 commented:
The specific choice of the `##()` syntax is because this is how it is in Dolphin Smalltalk. I work with an application that runs in both dialects, so I am particularly sensitive to the value of maintaining compatibility where possible—though if a clearly superior solution exists, I'm not against divergence. I don't think this is such a case though.

Your idea of an extensible special-expression syntax is neat, but IMO it's a poor tradeoff:
* Many of the use-cases here must be concise to be worthwhile, so including a long keyword would defeat the point.
* I think it's a case of YAGNI (You Ain't Gonna Need It)—I think it's far more likely in 10 years that nobody will have implemented another `##somethingelse()` than that someone will want a `###()` or similar.
* You're right that a keyword would be more explicit about what it does, but Smalltalk syntax remains simpler than most other popular languages even including more modern syntax extensions like brace-arrays and this compile-time syntax—I don't think it's a large burden for someone to look at a syntax cheatsheet the first time they encounter this.
* I personally find the use of english words as part of syntax to be an ugly departure from the rest of Smalltalk syntax, which IMO strikes a beautiful balance of keywords that are _de facto_ syntax (messages like `ifTrue:ifFalse:`, `not`, `whileTrue[:]`, `isNil`, and even `do:`) and punctuation that provides just enough structure for everything to hang together. I don't mind e.g. Ruby, Python, Lua's use of words in place of braces, but it's a pretty different design philosophy.

To answer your questions about bindings:
* The spec _does_ explicitly state that `self` is always the class, and specifically the instance class, never the metaclass. I added notes about the behavior in a workspace or debugger, and about class variables.
* `thisContext` is fundamentally bound by the VM to the actual current execution context, and inside a CTE is no different—it is being executed, after all. One _could_ use stack introspection to examine e.g. the actions that led to the method being compiled, but frankly, anyone who does this deserves the brittleness of the resulting code.

The uses-cases as I'm aware of them are quite well explained in the PhEP (see the section "Examples"), and as it says, the most common in my experience is literal dictionaries. I added a note that it is clearly bad for such dictionaries to be mutable—the use-case I'm talking about is not one of global mutable state but simply constant values that happen to be dictionaries rather than e.g. strings. Often they are very much the opposite of global, as well—they never leave the method in which they are declared! I almost certainly have a relatively narrow view of this, and I'd be curious to hear what use-cases others might come up with, but I do think this is **a** valid use-case. Your other concerns are more reasonable, it's just that honestly, if literal dictionaries end up being the only use-case, I would consider that justification enough. Specifically:

* I agree, `##(self)` feels a little "dirty" in some ways. I'm not terribly attached to it—if the consensus is that this is code smell and it's better to be explicit with a name, I would be fine with a lint rule to that effect. But I'm not sure explicitly naming the class is better, and indeed there already _is_ a lint rule for "Refers to class name instead of `self class`"—this is an alternative for the cases where the static binding is necessary and changing to `self class` would produce the wrong behavior.
* Having the compiler optimize constant sub-expressions like `60 * 60 * 24` could be reasonable, yes, and indeed the VM may already do so if the method triggers optimization in the JIT. This is more relevant in Dolphin, which does not have a JIT.
